### ISSUE TITLE
Documentation request: Why do Task names get reallocated into a Box<str> ?

### ISSUE THREAD

**Issue by rw** [async-rs/async-std#638]
When a String is provided as a Task name, it always gets reallocated (AFAICT) in this line of code:

https://github.com/async-rs/async-std/blob/master/src/task/task.rs#L33
```
            name: name.map(String::into_boxed_str),
```

For my use case, this forces Task names to be unusable because that causes a lot of heap churn. If there's a good reason for this allocation, then I recommend documenting it. Thanks!

**Comment by sdroege** [comment]
From following the code, `String::into_boxed_str()` -> `Vec::into_boxed_slice()` -> `RawVec::into_box()` -> `Box::from_raw()` don't do any allocations at all and only wrap the already heap-allocated memory of the `String` in a different way.

Maybe the initial `String` allocations are your problem? Do you have a testcase that can show the problem more accurately?

**Comment by rw** [comment]
@sdroege I think `shrink_to_fit` does reallocate, no?

**Comment by sdroege** [comment]
Indeed, but that depends then on the actual allocator in use here. I would expect that any sensible allocator would not allocate allocate new memory, copy data over and then free the old memory but instead would only mark the "unneeded" part at the end of the `Vec` as free in one way or another.

**Comment by rw** [comment]
@sdroege That's interesting, but relying on allocator behavior like that seems brittle. Are you saying you think mandating a `String` argument is a good idea, or are you only debating about the practicality of me worrying about these kinds of allocations?

It seems like the `name` code and data aren't defined in a performance-sensitive way, which is my fundamental concern. I'd be happy to experiment with updating it to be zero-alloc, if the project maintainers are open to that.

**Comment by sdroege** [comment]
> Are you saying you think mandating a `String` argument is a good idea, or are you only debating about the practicality of me worrying about these kinds of allocations?

Neither of the two, just an idle comment. IMHO requiring another allocation just for the name of the task is rather suboptimal, but having some testcase to show its effect in practice would be useful. Also for showing how much any alternative strategies improve performance in that testcase.

**Comment by rw** [comment]
@sdroege In my admittedly finite experience, showing the effect of heap allocations is very situation-dependent. Because the heap is usually a ginormous global variable shared by the whole program, you can get "spooky-action-at-a-distance" effects. For the same reason, you can get "spooky-no-action-from-any-distance" non-effects. Put more simply, I think it's hard to show if a particular heap allocation is particularly bad, because it is so use-case dependent.

In my case, I'm finding that I'm continually surprised that the main types in async-std are not traits. I really like customizing things e.g. for running many tests quickly by removing the overhead of heap allocations and atomic integers and so on.

**Comment by yoshuawuyts** [comment]
Part of why we chose the API we did here is because `thread::Builder::name` uses the same interface: https://doc.rust-lang.org/std/thread/struct.Builder.html#method.name.

```rust
pub fn name(self, name: String) -> Builder;
```
